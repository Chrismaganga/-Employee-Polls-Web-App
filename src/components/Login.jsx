import { useState } from 'react';
import { connect } from 'react-redux';
import { handleLogin } from '../actions/shared';
import PropTypes from 'prop-types';
import { useNavigate } from 'react-router-dom';


const Login = (props) => {
  const navigate = useNavigate();
  const [formData, setFormData] = useState({
    username: '',
    password: '',
  });
  const [errors, setErrors] = useState({});
  const [loading, setLoading] = useState(false);

  const validateForm = () => {
    const newErrors = {};
    if (!formData.username) {
      newErrors.username = 'Username is required';
    }
    if (!formData.password) {
      newErrors.password = 'Password is required';
    }
    setErrors(newErrors);
    return Object.keys(newErrors).length === 0;
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    if (!validateForm()) return;

    setLoading(true);
    try {
    
      const { users } = props;
      const user = users[formData.username];
      
      if (!user || user.password !== formData.password) {
        throw new Error('Invalid credentials');
      }

      await props.dispatch(handleLogin(user.id));
      navigate('/');
    } catch {
      setErrors({ submit: 'Invalid username or password. Please try again.' });
    } finally {
      setLoading(false);
    }
  };

  const handleInputChange = (e) => {
    const { name, value } = e.target;
    setFormData(prev => ({
      ...prev,
      [name]: value
    }));
  };

  return (
    <div className="login-container">
      <div className="login-box">
        <h3 className="login-title">Please login to continue...</h3>
        <h1 className="login-title">Employee Polls</h1>

        <form onSubmit={handleSubmit} className="login">
          <div className="form-group">
            <input
              type="text"
              id="username"
              name="username"
              value={formData.username}
              onChange={handleInputChange}
              className={errors.username ? 'error' : ''}
              placeholder="Enter your username"
              autoComplete="username" />
            {errors.username && <span className="error-message">{errors.username}</span>}
          </div>

          <div className="form-group">
            <input
              type="password"
              id="password"
              name="password"
              value={formData.password}
              onChange={handleInputChange}
              className={errors.password ? 'error' : ''}
              placeholder="Enter your password"
              autoComplete="current-password" />
            {errors.password && <span className="error-message">{errors.password}</span>}
          </div>
          
          {errors.submit && (
            <div className="error-message text-center">{errors.submit}</div>
          )}

          <button
            type="submit"
            className={`submit-button ${loading ? 'loading' : ''}`}
            disabled={loading}
          >
            {loading ? 'Signing in...' : 'Sign In'}
          </button>
        </form>
        <div className="login-footer">
          <p>Available Users:</p>
          <ul className="users-list">
            <li>johndoe / pass123</li>
            <li>janedoe / pass456</li>
          </ul>
        </div>
      </div>
    </div>
  );
};

Login.propTypes = {
  dispatch: PropTypes.func.isRequired,
  users: PropTypes.object.isRequired,
};

const mapStateToProps = ({ users }) => ({
  users,
});

export default connect(mapStateToProps)(Login);