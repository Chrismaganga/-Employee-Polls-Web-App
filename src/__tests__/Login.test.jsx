import { render, screen } from '@testing-library/react';
import { expect, describe, test } from '@jest/globals';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import Login from '../components/Login';

const mockStore = configureStore([]);

describe('Login Component', () => {
  test('renders login form', () => {
    const store = mockStore({
      users: {}
    });

    render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      </Provider>
    );

    // Check for the presence of key elements
    expect(screen.getByText('Employee Polls')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Enter your username')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const store = mockStore({
      users: {}
    });

    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      </Provider>
    );
    
    expect(container).toMatchSnapshot();
  });
});