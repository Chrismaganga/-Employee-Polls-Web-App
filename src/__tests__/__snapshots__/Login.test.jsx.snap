// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component matches snapshot 1`] = `
<div>
  <div
    class="login-container"
  >
    <div
      class="login-box"
    >
      <h3
        class="login-title"
      >
        Please login to continue...
      </h3>
      <h1
        class="login-title"
      >
        Employee Polls
      </h1>
      <form
        class="login"
      >
        <div
          class="form-group"
        >
          <input
            autocomplete="username"
            class=""
            id="username"
            name="username"
            placeholder="Enter your username"
            type="text"
            value="johndoe"
          />
        </div>
        <div
          class="form-group"
        >
          <input
            autocomplete="current-password"
            class=""
            id="password"
            name="password"
            placeholder="Enter your password"
            type="password"
            value="pass123"
          />
        </div>
        <button
          class="submit-button "
          type="submit"
        >
          Sign In
        </button>
      </form>
      <div
        class="login-footer"
      />
    </div>
  </div>
</div>
`;
